File: 2015/Day1/src/day1.cpp
#include <iostream>
#include <fstream>

using namespace std;
int main(int argc, char * argv[])
{
    if(argc != 2)
    {
        cout << "usage: ./day1 input.txt" << endl;
        exit(1);
    }
    
    int floor = 0;
    unsigned int command_counter = 0;
    unsigned int basement_position = 0;
    char command;
    ifstream file(argv[1]);

    // read from the input file character by character to keep
    // track of Santa's floor
    while(file >> command)
    {
        command_counter += 1;
        if(command == '(')
        {
            floor += 1;
        }
        else if(command == ')')
        {
            floor -= 1;
        }
        if(basement_position == 0 && floor < 0) //only set basement_position once
        {
            basement_position = command_counter;
        }
   }
   file.close();

   cout << "Part 1 answer is " << floor << endl;
   cout << "Part 2 answer is " << basement_position << endl; 
}

File: 2015/Day1/src/day1.py
import sys

def part1(line: str) -> int:
    floor = 0
    for char in line:
        if char == "(":
            floor += 1
        elif char == ")":
            floor -= 1
    return floor

def part2(line: str) -> int:
    floor = 0
    for i in range(len(line)):
        char = line[i]
        if char == "(":
            floor += 1
        elif char == ")":
            floor -= 1
        if floor < 0:
            return (i + 1)
    return len(line) # floor was never negative

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("usage: python3 day1.py input.txt")
        exit(1)

    line = ""
    with open(sys.argv[1], "r") as fileobj:
        line = fileobj.readline().strip()

    part1_ans = part1(line)
    part2_ans = part2(line)
    
    print(f"Part 1: Santa ends up on floor {part1_ans}")
    print(f"Part 2: position {part2_ans} causes Santa to enter basement")
